engine/runtime/src/main/java/org/enso/interpreter/caches/Cache.java

```diff
@@ -318,11 +305,11 @@ private Optional<T> loadCacheFrom(
    * @param path location of the serialized metadata
    * @return deserialized metadata, or [[None]] if invalid
    */
-  private Optional<M> loadCacheMetadata(TruffleFile path, TruffleLogger logger) throws IOException {
+  private M loadCacheMetadata(TruffleFile path, TruffleLogger logger) throws IOException {
```

Comment:
The problem with `Optional` is that it doesn't specify reason why it is `Optional.empty()`. To track the `HelloWorldCacheTest` random CI failures down, we'd rather understand the reasons why a cache file isn't loaded, thus:
- returning _non-null_ value from the method or
- throwing `IOException` with a message describing the reason why we cannot load the cache file

Hopefully this uncovers the reason why the `HelloWorldCacheTest` randomly fails.

Reply:
That's pretty much a definition of `Either`

Reply:
Right, that is true. I read an article arguing in favor of `Either` over `Option` as the former can provide reason for the failure. However we are in Java and I believe the Java equivalent of `Either` are [checked exceptions](http://wiki.apidesign.org/wiki/Either).

Reply:
- technically we need just a6b67e4c469f2d985d65a75c249f3e25ebc7820b to fix #11196 
- Java doesn't have `Either` as far as I know
- rewriting to `scala.Either` goes opposite direction than we are moving to
- I can back out 691136dc6cb5924a3c86ba60e3f6fd75dd8b5d65 to smooth the review process

Vote +1 if you want to me to back 691136dc6cb5924a3c86ba60e3f6fd75dd8b5d65 out, but you approve the rest of changes. Thank you.

Reply:
Done as:
```diff
-   * @return the cached data if available, otherwise an empty [[Optional]].
+   * @return the cached data if available, otherwise throw exception
```

Reply:
Fair enough. We are in Java and Java is no friends with FP.